### PR TITLE
Improve performance and memory usage during report file generation

### PR DESCRIPTION
**Pull request by alexanderkozlenko**
The PR improves performance and memory usage during report file generation by the following changes:

1. Remove additional allocation of  `MemoryStream`, `List<T>`, `byte[]`, and `string` objects.
2. Change reporters to write content directly to a `StreamWriter` instance.

### CODE REVIEW

File: src/coverlet.console/Program.cs [hunk 1]
```diff
@@ -77,7 +77,9 @@ static int Main(string[] args)
 
                     var report = Path.Combine(directory, filename);
                     logger.LogInformation($"  Generating report '{report}'");
-                    File.WriteAllText(report, reporter.Report(result));
+                    using (var fileStream = new FileStream(report, FileMode.Create, FileAccess.Write, FileShare.Read))
```
Comment (latop2604) [inline src/coverlet.console/Program.cs]: Why not just `var streamWriter = File.CreateText(...)` ?

**Comment by joopscheer** [comment]
This PR is missing Codecov results. Any idea why?

**Comment by alexanderkozlenko** [comment]
@joopscheer, I'm not sure that this is what you are asking about, however, this is code coverage report from AppVeyor for the PR:

```
  +------------------+--------+--------+--------+
  | Module           | Line   | Branch | Method |
  +------------------+--------+--------+--------+
  | coverlet.core    | 91.4%  | 86.3%  | 93.3%  |
  +------------------+--------+--------+--------+
  | coverlet.tracker | 0%     | 0%     | 0%     |
  +------------------+--------+--------+--------+
```
and this is the latest code coverage result for the master branch:
```
  +------------------+--------+--------+--------+
  | Module           | Line   | Branch | Method |
  +------------------+--------+--------+--------+
  | coverlet.core    | 91.4%  | 86.3%  | 93.2%  |
  +------------------+--------+--------+--------+
  | coverlet.tracker | 0%     | 0%     | 0%     |
  +------------------+--------+--------+--------+
```

**Comment by joopscheer** [comment]
@alexanderkozlenko Never mind. Some PR's have four checks. This PR just one. Not sure why. But that's why there isn't a Codecov report comment.

**Comment by tonerdo** [comment]
Hi @alexanderkozlenko, thanks for the contribution. Can you include some stats on the performance improvement. You can use `coverlet.core.performancetest` as a baseline.

**Comment by tonerdo** [comment]
@alexanderkozlenko, that's one way to go. Specifically the `coverlet.tests.performancetest` project

**Comment by alexanderkozlenko** [comment]
Unfortunately, I cannot provide with meaningful performance testing results due to the following reasons:

1. The main change is reducing allocations, that is not reflected in test runner output.
2. The accuracy of the time results is not so good to be considered as a performance benchmark.

I assume the only appropriate way to measure performance changes is to use some performance benchmark framework (like [BenchmarkDotNet](https://benchmarkdotnet.org)). It seems it also requires some additional changes to the source code to allow easy mocking of coverage results.

**Comment by tonerdo** [comment]
Hmmm... perhaps @petli can advise in this situation

**Comment by petli** [comment]
The `coverlet.tests.performancetest` project as it is now is not suitable for testing the report generation, since there a lot of classes, methods and files are necessary. That performance test is only useful for testing the instrumentation and event collection. 

Another test could be added to it that mocks up a big coverage result and then measures the report generation time, like @alexanderkozlenko suggests.

**Comment by tonerdo** [comment]
Hi @alexanderkozlenko, any particular reason you closed this and your other PRs?

**Pull request closed without merging**